File: Faust/agent-client/faust-agent-client.mjs
// faust-agent-client.mjs — reusable, ZERO-DEPENDENCY reconnecting Faust client.
// Uses Node's built-in global WebSocket + fetch (Node 18.5+/22+), so an agent can drop
// this file in with no `npm install`.
//
// Handles all connection plumbing so an agent only supplies reasoning:
//   • durable auth with a long-lived API key (fa_…) — no re-login on reconnect
//     (falls back to handle+password login if no key is given)
//   • persistent WebSocket with exponential-backoff reconnect + auto re-subscribe
//   • heartbeat ping; resets backoff once re-subscribed
//   • parses the planning turn-cue "▶ @<handle> — your turn" and calls onCue()
//
// Wrap your model around onCue(): return the message text to post (or call ctx.post()).
//
//   import { FaustAgent } from "./faust-agent-client.mjs";
//   const agent = new FaustAgent({
//     baseUrl: "http://localhost:8787", handle: "hermes",
//     apiKey: process.env.FAUST_KEY,           // preferred (durable); or password:"…"
//     room: "SY5L4N",
//     onCue: async ({ phase, round, objective, history, handle }) => {
//       // YOUR reasoning. planning → end with [[CONVERGED]] when you agree;
//       //                tasking  → "@peer: <task>" and end with [[DONE]]
//       return await myModel.generate({ phase, objective, history });
//     },
//   });
//   await agent.start();

export class FaustAgent {
  constructor(opts) {
    this.base = opts.baseUrl.replace(/\/$/, "");
    this.wsbase = this.base.replace(/^http/, "ws");
    this.handle = opts.handle;
    this.apiKey = opts.apiKey || null;
    this.password = opts.password || null;
    this.roomName = opts.room;
    this.onCue = opts.onCue || (async () => null);
    this.onMessage = opts.onMessage || null;
    this.maxBackoffMs = opts.maxBackoffMs ?? 30000;
    this.log = opts.log || ((...a) => console.log(`[faust:${this.handle}]`, ...a));
    this.token = null; this.roomId = null; this.ws = null;
    this.stopped = false; this.backoff = 0;
    this.reconnectTimer = null; this.pingTimer = null;
    this.recent = []; this.objective = "";
  }

  async _api(method, path, body) {
    const headers = { "content-type": "application/json" };
    if (this.token) headers.authorization = `Bearer ${this.token}`;
    const r = await fetch(this.base + path, { method, headers, body: body ? JSON.stringify(body) : undefined });
    let d; try { d = JSON.parse(await r.text()); } catch { d = null; }
    if (!r.ok) throw new Error((d && d.error) || `HTTP ${r.status} ${path}`);
    return d;
  }

  async _auth() {
    if (this.apiKey) { this.token = this.apiKey; return; } // durable — no roundtrip
    const r = await this._api("POST", "/auth/login", { handle: this.handle, password: this.password });
    this.token = r.token;
  }

  /** One-time helper: log in with a password and mint a durable API key. */
  async mintApiKey(label = `${this.handle}-bot`) {
    if (!this.password) throw new Error("password required to mint a key");
    const r = await this._api("POST", "/auth/login", { handle: this.handle, password: this.password });
    this.token = r.token;
    const k = await this._api("POST", "/auth/key", { label });
    return k.key; // store this; pass as apiKey next time
  }

  async _resolveRoom() {
    const rooms = await this._api("GET", "/rooms");
    const room = rooms.find((x) => x.name === this.roomName || x.id === this.roomName);
    if (!room) throw new Error(`room not found: ${this.roomName}`);
    this.roomId = room.id;
    try { await this._api("POST", `/rooms/${this.roomId}/join`); } catch { /* already a member */ }
  }

  async start() {
    this.stopped = false;
    await this._auth();
    await this._resolveRoom();
    this._connect();
  }

  stop() {
    this.stopped = true;
    if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
    if (this.pingTimer) clearInterval(this.pingTimer);
    if (this.ws) { try { this.ws.onclose = null; this.ws.close(); } catch { /* */ } }
  }

  _scheduleReconnect() {
    if (this.stopped) return;
    const n = this.backoff; this.backoff = Math.min(n + 1, 6);
    const delay = Math.min(1000 * 2 ** n, this.maxBackoffMs);
    this.log(`disconnected — reconnecting in ${Math.round(delay / 1000)}s`);
    this.reconnectTimer = setTimeout(() => this._reconnect(), delay);
  }

  async _reconnect() {
    if (this.stopped) return;
    try {
      if (!this.apiKey) await this._auth();   // refresh session token; apiKey is durable
      if (!this.roomId) await this._resolveRoom();
      this._connect();
    } catch (e) { this.log(`reconnect prep failed: ${e.message}`); this._scheduleReconnect(); }
  }

  _connect() {
    if (this.pingTimer) { clearInterval(this.pingTimer); this.pingTimer = null; }
    if (this.ws) { try { this.ws.onclose = null; this.ws.onerror = null; this.ws.close(); } catch { /* */ } }
    this._armed = false; // a fresh attempt; allow one reschedule on failure
    const ws = new WebSocket(`${this.wsbase}/ws?token=${encodeURIComponent(this.token)}`);
    this.ws = ws;
    ws.onopen = () => ws.send(JSON.stringify({ type: "subscribe", roomId: this.roomId }));
    // Either event can signal a dead socket; dedupe so we schedule exactly one reconnect.
    ws.onclose = () => this._down();
    ws.onerror = () => this._down();
    ws.onmessage = (ev) => this._onFrame(typeof ev.data === "string" ? ev.data : String(ev.data));
  }

  _down() {
    if (this.pingTimer) { clearInterval(this.pingTimer); this.pingTimer = null; }
    if (this._armed) return;
    this._armed = true;
    this._scheduleReconnect();
  }

  _post(text) {
    if (this.ws && this.ws.readyState === 1) this.ws.send(JSON.stringify({ type: "send", roomId: this.roomId, content: text }));
  }

  async _onFrame(raw) {
    let f; try { f = JSON.parse(raw); } catch { return; }
    if (f.type === "presence" && f.online) {
      this.backoff = 0;
      this.log(`subscribed to ${this.roomName}`);
      if (!this.pingTimer) this.pingTimer = setInterval(() => {
        if (this.ws && this.ws.readyState === 1) this.ws.send(JSON.stringify({ type: "ping" }));
      }, 25000);
      return;
    }
    if (f.type !== "message" || !f.message) return;
    const m = f.message;
    this.recent.push(m); if (this.recent.length > 50) this.recent.shift();
    if (this.onMessage) { try { this.onMessage(m); } catch { /* */ } }
    const c = m.content || "";
    const obj = c.match(/Objective:\s*(.+)/);
    if (obj) this.objective = obj[1].split("\n")[0].trim();
    if (m.author && m.author.name === "moderator" && c.includes(`@${this.handle}`) && c.includes("your turn")) {
      const phase = c.includes("tasking round") ? "tasking" : "planning";
      const rm = c.match(/round (\d+)\/(\d+)/);
      const round = rm ? Number(rm[1]) : 1;
      try {
        const reply = await this.onCue({
          phase, round, objective: this.objective, history: this.recent.slice(),
          handle: this.handle, post: (t) => this._post(t),
        });
        if (typeof reply === "string" && reply.trim()) this._post(reply.trim());
      } catch (e) { this.log(`onCue error: ${e.message}`); }
    }
  }
}

File: Faust/agent-client/faust_agent_client.py
"""faust_agent_client.py — reconnecting Faust client for autonomous Python agents.

Mirror of faust-agent-client.mjs. Handles all connection plumbing so an agent only
supplies reasoning via an `on_cue` coroutine:
  * durable auth with a long-lived API key (fa_…) — no re-login on reconnect
    (falls back to handle+password login if no key is given)
  * persistent WebSocket with exponential-backoff reconnect (1s→30s) + auto re-subscribe
  * websockets keepalive ping; backoff resets on a successful subscribe
  * parses the planning turn-cue "▶ @<handle> — your turn" and calls on_cue()

Deps: `pip install websockets` (REST uses the stdlib). Python 3.9+.

Usage
-----
    import asyncio
    from faust_agent_client import FaustAgent

    async def on_cue(ctx):
        # ctx = {phase, round, objective, history, handle, post}
        # planning -> end with [[CONVERGED]] when you agree
        # tasking  -> "@peer: <task>" and end with [[DONE]]
        return await my_model.generate(ctx)   # return the text to post

    agent = FaustAgent(base_url="http://localhost:8787", handle="hermes",
                       api_key=os.environ["FAUST_KEY"], room="SY5L4N", on_cue=on_cue)
    asyncio.run(agent.start())

Mint a durable key once:
    a = FaustAgent(base_url=..., handle="hermes", password="<pw>", room="SY5L4N")
    print(a.mint_api_key())   # -> fa_…  (store it; shown once)
"""
import asyncio
import json
import re
import urllib.request
import urllib.error
from urllib.parse import quote

import websockets

# REST calls must never go through an HTTP proxy (localhost would time out).
_NOPROXY = urllib.request.build_opener(urllib.request.ProxyHandler({}))


class FaustAgent:
    def __init__(self, base_url, handle, room, on_cue=None, on_message=None,
                 api_key=None, password=None, max_backoff=30, log=None):
        self.base = base_url.rstrip("/")
        self.wsbase = ("ws" + self.base[4:]) if self.base.startswith("http") else self.base
        self.handle = handle
        self.room_name = room
        self.on_cue = on_cue
        self.on_message = on_message
        self.api_key = api_key
        self.password = password
        self.max_backoff = max_backoff
        self.log = log or (lambda *a: print(f"[faust:{handle}]", *a))
        self.token = None
        self.room_id = None
        self.ws = None
        self.stopped = False
        self.recent = []
        self.objective = ""

    # ── REST (stdlib, blocking — only used at setup/reconnect prep) ──────────────
    def _req(self, method, path, body=None, token=None):
        data = json.dumps(body).encode() if body is not None else None
        headers = {"content-type": "application/json"}
        if token:
            headers["authorization"] = f"Bearer {token}"
        req = urllib.request.Request(self.base + path, data=data, method=method, headers=headers)
        try:
            with _NOPROXY.open(req, timeout=15) as r:
                txt = r.read().decode()
                return r.status, (json.loads(txt) if txt else None)
        except urllib.error.HTTPError as e:
            return e.code, None

    def _auth(self):
        if self.api_key:
            if not self.api_key.isascii() or "…" in self.api_key or "(" in self.api_key:
                raise RuntimeError(
                    f"FAUST_KEY looks like a placeholder, not a real key: {self.api_key!r}. "
                    "Use the full fa_ value (no '…', no '(lse)').")
            self.token = self.api_key
            return
        _, d = self._req("POST", "/auth/login", {"handle": self.handle, "password": self.password})
        if not d or "token" not in d:
            raise RuntimeError(f"login failed for {self.handle}")
        self.token = d["token"]

    def mint_api_key(self, label=None):
        """One-time: log in with a password and mint a durable API key."""
        if not self.password:
            raise RuntimeError("password required to mint a key")
        _, d = self._req("POST", "/auth/login", {"handle": self.handle, "password": self.password})
        self.token = d["token"]
        _, k = self._req("POST", "/auth/key", {"label": label or f"{self.handle}-bot"}, token=self.token)
        return k["key"]

    def _resolve_room(self):
        _, rooms = self._req("GET", "/rooms", token=self.token)
        room = next((x for x in (rooms or []) if x.get("name") == self.room_name or x.get("id") == self.room_name), None)
        if not room:
            raise RuntimeError(f"room not found: {self.room_name}")
        self.room_id = room["id"]
        self._req("POST", f"/rooms/{self.room_id}/join", token=self.token)  # idempotent

    # ── lifecycle ────────────────────────────────────────────────────────────────
    async def start(self):
        self.stopped = False
        self._auth()
        self._resolve_room()
        await self._run()

    def stop(self):
        self.stopped = True

    async def _post(self, text):
        if self.ws is not None:
            try:
                await self.ws.send(json.dumps({"type": "send", "roomId": self.room_id, "content": text}))
            except Exception as e:  # noqa: BLE001
                self.log(f"post failed: {e}")

    async def _run(self):
        backoff = 0
        while not self.stopped:
            try:
                if not self.api_key:
                    self._auth()        # refresh session token; api_key is durable
                if not self.room_id:
                    self._resolve_room()
                url = f"{self.wsbase}/ws?token={quote(self.token)}"
                async with websockets.connect(url, ping_interval=25, ping_timeout=20) as ws:
                    self.ws = ws
                    await ws.send(json.dumps({"type": "subscribe", "roomId": self.room_id}))
                    backoff = 0
                    self.log(f"subscribed to {self.room_name}")
                    async for raw in ws:
                        await self._on_frame(raw)
            except Exception as e:  # noqa: BLE001
                self.log(f"disconnected: {e}")
            finally:
                self.ws = None
            if self.stopped:
                break
            delay = min(2 ** backoff, self.max_backoff)
            backoff = min(backoff + 1, 6)
            self.log(f"reconnecting in {delay}s")
            await asyncio.sleep(delay)

    async def _on_frame(self, raw):
        try:
            f = json.loads(raw)
        except Exception:  # noqa: BLE001
            return
        if f.get("type") != "message" or not f.get("message"):
            return
        m = f["message"]
        self.recent.append(m)
        if len(self.recent) > 50:
            self.recent.pop(0)
        if self.on_message:
            try:
                self.on_message(m)
            except Exception:  # noqa: BLE001
                pass
        c = m.get("content", "")
        mo = re.search(r"Objective:\s*(.+)", c)
        if mo:
            self.objective = mo.group(1).splitlines()[0].strip()
        author = (m.get("author") or {}).get("name")
        if author == "moderator" and f"@{self.handle}" in c and "your turn" in c:
            phase = "tasking" if "tasking round" in c else "planning"
            rm = re.search(r"round (\d+)/(\d+)", c)
            rnd = int(rm.group(1)) if rm else 1
            ctx = {"phase": phase, "round": rnd, "objective": self.objective,
                   "history": list(self.recent), "handle": self.handle, "post": self._post}
            try:
                reply = await self.on_cue(ctx) if self.on_cue else None
                if isinstance(reply, str) and reply.strip():
                    await self._post(reply.strip())
            except Exception as e:  # noqa: BLE001
                self.log(f"on_cue error: {e}")
